### DISCUSSION TITLE
Is it possible to re-use spack installed packages inside another environment

### DISCUSSION THREAD

**Discussion by TomMelt** [spack/spack#35086] [General]
I have seen the command `spack external find` generates a `packages.yaml` file for external packages.

When I generate a new spack environment using `spack env create` I sometimes want to use packages that already exist in my root spack installation.

I want the new environment to pickup packages/specs that are already installed in root and only add new ones to the new environment. 

I thought therefore, I could use a `packages.yaml` file to add the root packages as externals.

I have tried using `spack add spec` but it doesn't work as expected and I would have to manually add a lot of specs in some cases. 

**tldr Question:**  Is this possible to create a `packages.yaml` file for spack installed programs/libs or is it a bad idea?

Here is an example of what I want using a clean install in the docker image:

```bash
docker pull spack/ubuntu-jammy 
docker run -it spack/ubuntu-jammy
```

As an example, I generate my root install to have all the prerequisites for `json-fortran` (as an example)
```bash
spack install json-fortran
spack uninstall -y json-fortran
spack find -lN
```

Now if I create a new spack env I can install json-fortran and it uses all my currently installed packages (but, `json-fortran` is added to my main install path: `/opt/spack/opt/spack/`
```bash
spack env create -d ./test
spack -e ./test/ add json-fortran
spack -e ./test/ install
```

However, if I install into a different `install_tree` then it re-installs all the dependencies again.
```bash
spack uninstall -y json-fortran
spack -e ./test/ config add "config:install_tree:root:/root/test/opt/spack"
spack -e ./test/ install
```

**Comment by TomMelt** [comment]
This discussion can be "closed". I was just a bit confused with the eco-system. I now know how to achieve what I want using the `$HOME/.spack/packages.yaml`  config file.

**Comment by BeauVerrill7Q** [comment]
Yes, it is possible to create a packages.yaml file for Spack installed packages/libs. The spack external find command can be used to generate a packages.yaml file for external packages, but it won't include packages already installed in your root Spack installation.

One way to achieve what you want is to use spack find --format "{name}@{version}" to generate a list of all the packages installed in your root Spack installation, and then manually add them to the packages.yaml file for your new environment.

Here are the steps to do this:

Generate a list of installed packages in your root Spack installation:
spack find --format "{name}@{version}" > installed_packages.txt

Copy the packages.yaml file from your new environment directory (./test) to a temporary directory, and open it in a text editor:
cp ./test/packages.yaml ./temp_packages.yaml
nano ./temp_packages.yaml

Add the list of installed packages to the packages.yaml file under the specs key:
specs:
  - json-fortran@20211117
  - zlib@1.2.11
  - ...
  - ...
  Save and close the temp_packages.yaml file.

Use the spack env activate command to activate your new environment, and then install the json-fortran package:
spack env activate ./test/
spack install json-fortran

Spack should now use the packages already installed in your root Spack installation, and only install json-fortran in your new environment.

Note that if the package versions in your new environment are different from those in your root Spack installation, you may need to adjust the specs list in the packages.yaml file accordingly.

**Reply by TomMelt** [reply]
Did you mean like this (see below). If so this still tries to re-install all the packages in the second example (`spack -e ./test/ config add "config:install_tree:root:/root/test/opt/spack"`)

```
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs: [json-fortran,
    berkeley-db@18.1.40,
    bzip2@1.0.8,
    ca-certificates-mozilla@2023-01-10,
    cmake@3.25.1,
    diffutils@3.8,
    gdbm@1.23,
    ncurses@6.4,
    openssl@1.1.1s,
    perl@5.36.0,
    pkgconf@1.8.0,
    readline@8.2,
    zlib@1.2.13]
  view: true
  concretizer:
    unify: true
  config:
    install_tree:
      root: /root/test/opt/spack
```

**Comment by TomMelt** [comment]
Thanks a lot for the replies. This is really helpful

**Comment by KelleyKjeldgaardW** [comment]
It is possible to create a packages.yaml file for spack installed programs/libs, but it may not be the best solution. One way to achieve this would be to use the spack external find command to generate a packages.yaml file for the desired external packages and then use that packages.yaml file in the spack environment creation process. However, this may not be the most efficient way to accomplish the goal of adding root packages as externals to a new environment.

A better approach may be to use the spack mirror create command to create a mirror of the root install and then use that mirror as a source for the new environment. This will avoid the need to re-install dependencies that are already present in the root installation.

For example, to create a mirror of the root install, run:
![image](https://user-images.githubusercontent.com/127733138/224658727-77b0ac66-7cf1-420a-8520-df0ac3c7f919.png)
Then, when creating the new environment, specify the mirror as a source:
![image](https://user-images.githubusercontent.com/127733138/224658796-c9ebcd32-b514-4bf3-8bb7-5830f364ddb6.png)
This will use the packages from the mirror, avoiding the need to re-install them, and only install the new package in the new environment.
